elkscmd/man/man8/makeboot.8
.TH MAKEBOOT 8
.SH NAME
makeboot \- prepare device as a system boot device
.SH SYNOPSIS
\fBmakeboot\fP [\fB\-M|F\fP] [\fB\-s\fP] [\fB\-f file\fR] \fIdevice\fR
.br
.SH OPTIONS
.TP 5
.B \-M
Write the ELKS MBR (Master Boot Record) to the first sector of the specified device.
.TP 5
.B \-F
Create a 'flat' (unpartitioned) device. 
.TP 5
.B \-f file
Take the bootblock from the specified instead of the current root device.
.TP 5
.B \-s
Copy system files from the current root to the target device.
.SH EXAMPLES
.TP 5
Copy the bootblock from the current root partition to \fB/dev/hda4\fP:
.sp
.nf
# makeboot /dev/hda4
System on /dev/hda1: Minix (CHS 7818/16/63 at offset 63)
Target on /dev/hda4: Minix (CHS 7818/16/63 at offset 7560000)
Bootblock written
.fi
.TP 5
Write the ELKS MBR til the specified device:
.sp
.nf
# makeboot \-M /dev/hda4
Writing MBR on /dev/hda
.fi
.sp
Notice that while a partition (/dev/hda4) is specified on the command line, 
.B makeboot
changes that into a device (/dev/hda) since that's where the MBR goes.
.TP 5
Replace the bootblock on \fB/dev/hda4\fP with the one in \fB./minix.bin\fP and copy system files to the target partition.
.sp
.nf
# makeboot \-s \-f ./minix.bin /dev/hda4
Target on /dev/hda4: Minix (CHS 7818/16/63 at offset 7560000)
Bootblock written
Copying /linux to /tmp/mnt/linux
Copying /bootopts to /tmp/mnt/bootopts
System copied
.fi
.SH DESCRIPTION
\fBmakeboot\fR prepares a device or partition for booting by installing a 
bootloader and \- if the \fB-s\fP option is present \- some system files.
The partition or device must already contain a file system, which may be either FAT (MSDOS) 
or Minix.
.PP
Without options, a bootblock is copied from the current root device to the target. 
A different bootblock may be specifying via the \fB-f\fP option, see examples above.
Such transfer requires that the running system and the target has the same file system types.
.B makeboot
will complain and exit if this is not the case.
.PP
If the 
.B -s
option is present, the ELKS kernel (\fI/linux\fR) and the kernel configuration file 
.I /bootopts
will be copied to the target.
.PP
The 
.B -M
option may be used to write a Master Boot Record (MBR) to the target device. The 
MBR will preserve the partition information currently present in the first sector of the target device.
This operation is similar to the MSDOS 
'fdisk /MBR' command.
Refer to the
.BR MBR (8)
manpage for details about the ELKS MBR.
.PP
.B makeboot
also supports 'flat devices' - aka unpartitioned devices. While lloppies are always unpartitioned,
hard sisks usually have partitions and a Master Boot Record. By writing the 
.I bootblock 
to the first sector(s) of the disk, the entire disk becomes a single partition. The
.I -F
option does this.
.PP
.B makeboot
is used by the 
.B sys
command to build a fully pouplated and bootable ELKS system from floppy or from another disk.
.PP
\fBfloppy\fP:
on the specified devices.
.SH BUGS
.B makeboot
cannot replace the MBR on the current root device.
.SH "SEE ALSO"
.BR sys (8),
.BR mkfs (8),
.BR minix (8),
.BR FAT (8),
.BR MBR (8),
.BR boot (8).


elks/CHANGELOG
This file is used to record changes made to the CVS database. Please add to
the top of this file before committing any changes. Please use the header
format provided for each entry, noting that the text on each line should not
be longer than the lines of dashes.

Al Riddoch.

Contributor           Email address                     CHANGELOG entries found
~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~
Alan Cox              alan@redhat.com                   17 Jun 2001-18 Jun 2001
Alastair Bridgewater  nyef@softhome.net                 19 Nov 1998- 6 Dec 1998
Alistair Riddoch      alriddoch@zepler.org              25 Sep 1998-17 Jun 2001
Arnaud Launay         asl@launay.org                    15 Jul 1999-17 Oct 2001
Christian Mardmuller  chm@kdt.de                        10 Nov 1999-29 Feb 2000
Christoph Plattner    Christoph.Plattner@dot.at         31 Jan 2000- 2 Feb 2000
Claudio Matsuoka      claudio@conectiva.com
Daniele Giacomini     daniele@swlibero.org		20 Sep 2004
David Murn            scuffer@hups.apana.org.au         11 Nov 1998-15 Nov 1999
Eduardo Habkost	      ehabkost@conectiva.com.br		01 Jun 2004
Giles Russell                                           22 Nov 1999
Greg Haerr                                              14 Jul 1999-20 Jul 1999
Harry Kalogirou       HarKal@gmx.net                    20 Jun 2001- 2 Jan 2003
Jacek Lipkowski       sq5bpf@acid.ch.pw.edu.pl          20 Jan 2002
Jay Barr                                                27 Oct 2002
Ken Martwick          kenm@efn.org                      12 May 2002
M Warner Losh         imp@village.org                   20 Jan 2002
Manuel Novoa III      mjn3@codepoet.org                 18 Jun 2002-25 Jun 2002
Mario Frasca                                            20 Jan 2000-14 Feb 2000
Miguel Bolanos	      mike@linuxlabs.com		20 Sep 2004
Paul Nasrat           pauln@truemesh.com                29 Oct 2001-12 Jul 2002
Per Olofsson          MagerValp@cling.gu.se             25 Feb 2000
Peter L Vachuska      pvachusk@nconnect.net             13 Aug 2001
Richard Wallman       murkygoth@murkygoth.uklinux.net    2 Apr 2002
Riley Williams        Riley@Williams.Name               16 Nov 1998-13 Jul 2003
Roman Festchook       roma@polesye.net                  27 Oct 2001
Simon Weijgers        simon@weijgers.com                16 Sep 1998-15 Nov 1999
Simon Wood                                               6 Jan 2000-12 Jan 2000
Thomas McWilliams     tgm9999@yahoo.com                 27 Aug 1999- 2 Apr 2001
Tommy McCabe 	      rocketjet314@yahoo.com		31 May 2004

-------------------------------------------------------------------------------

Thu Feb 16 17:03:00 EST 2012	Jody Bruchon <jody@jodybruchon.com>	
	* Removed elksfs. It was a copy of the MINIX fs and no tools
	  exist to make such a filesystem.

Thu Nov 23 21:35:28 EST 2006	Segin <segin2005@gmail.com>
	* Changed name for the submenu for selecting charater
	  device drivers to note the fact that that submenu is broken;
	  attempting to enter it will have lxdialog die horribly.

Wed Nov  8 17:49:52 EST 2006	Segin <segin2005@gmail.com>

	* Changed ELKS init sequence to check all locations for 'init'
	  that the Linux kernel does. This is to follow more 
	  "traditional" behaviour.

Sun Aug 20 21:51:25 UTC 2006	Claudio Matsuoka <claudio@mandriva.com>

	* Allow loading Minix executables with total size larger than
	  64KB, such as ash (70140 bytes). Currently the total size
	  minus 32 bytes of header was limited to 16-bit.

	* Set tty1 back to minor 0. This is different from Linux
	  numbering, but it's necessary to have current ELKS virtual
	  terminals working. This and the executable size change put
	  ELKS back in usable state, allowing logins in the first VT
	  and executing a shell.

Mon Sep 20 23:27:23 CST 2004	Daniele Giacomini <daniele@swlibero.org>

	* Added italian keys support, Miguel Bolanos cleaned the original
	  file code.

Wed Jun  9 17:27:52 BRT 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* fix for my last fix: s/rp_cmd/rq_cmd/

Thu Jun 03 21:47:13 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* fixed format specifiers to correctly displaysegment sizes as unsigned
	  values, and to display debug messages correctly.

Thu Jun 03 07:48:25 CST 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* Fixed wrong struct field name for arch/i86/drivers/block/doshd.c
	* Several fixes ro arch/i86/mm/user.c

Web Jun 02 14:16:57 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* Added an extra fix to Eduardos memcpy fix.

Tue Jun 01 21:00:03 CST 2004	Eduardo Habkost ehabkost@conectiva.com.br

	* Fix memcpy_{to,from}fs, adding code to save the contents of SI and DI

Mon May 31 07:39:47 CST 2004	Tommy McCabe <rocketjet314@yahoo.com>

	* Fix wrong assignment (typo?) on printk
	* Fix order of LD parameters on arch/i86/Makefile
	* Use #ifdef rightly on setup.S
	* Fix init to use tty0, not tty1 as system console.

===============================================================================

Sun Jul 13 04:24:44 BST 2003	Riley Williams <Riley@Williams.Name>

	* This is ELKS 0.1.3-pre1

Sat Jul 12 12:24:36 BST 2003	Riley Williams <Riley@Williams.Name>

	* Rewrote kernel/printk.c to allow printk() to correctly print out
	  signed values.
	* Rewrote kernel/printk.c to produce a less verbose panic() dump,
	  and also to increase the number of lines of stack trace to 10.
	* Various spelling corrections to printk() messages.

===============================================================================

Thu Jan 02 16:12:36 EET 2003	Harry Kalogirou <harkal@gmx.net>

        * Better select hash.
	* This is ELKS 0.1.2

-------------------------------------------------------------------------------

Fri Nov 01 01:52:44 EET 2002	Harry Kalogirou <harkal@gmx.net>

        * Changed the position of ds in the _registers struct.
        * Added support for loading the new ELKS executable format. This is 

elkscmd/sys_utils/makeboot.c
/*
 * makeboot - make a bootable image
 * Part of /bin/sys script for creating ELKS images from ELKS
 *
 * Usage: makeboot [-M][-F] /dev/{fd0,fd1,hda1,hda2,etc}
 *
 * Copies boot sector from root device to target device
 * Sets EPB and BPB parameters in boot sector
 * Copies /linux and /bootopts for MINIX
 * Copies /linux and creates /dev for FAT
 * If -M: write MBR using compiled-in mbr.bin
 * If -F: allow writing to flat (non-MBR) hard drive
 * If -f file: take bootblock from file
 * If -s: Skip file copying, useful for bootblock copy only
 * If -m: MBR only, skip file copy and bootblock
 *
 * Nov 2020 Greg Haerr
 */

#include <stdio.h>
#include <dirent.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <fcntl.h>
#include <utime.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <sys/mount.h>
#include <arch/hdreg.h>
#include <linuxmt/minix_fs.h>
#include <linuxmt/kdev_t.h>
#include "../../bootblocks/mbr_autogen.c"

#define BUF_SIZE	1024 

#define MOUNTDIR	"/tmp/mnt"

#define SYSFILE1	"/linux"		/* copied for MINIX and FAT*/
#define SYSFILE2	"/bootopts"		/* copied for MINIX only */
#define DEVDIR		"/dev"			/* created for FAT only */

/* BIOS driver numbers must match bioshd.c*/
#define BIOS_NUM_MINOR	32		/* max minor devices per drive*/
#define BIOS_MINOR_MASK	(BIOS_NUM_MINOR - 1)
#define BIOS_FD0_MINOR	128		/* minor # of first floppy, must match bioshd.c*/

/* See bootblocks/minix.map for the offsets, these used for MINIX and FAT */
#define ELKS_BPB_NumTracks	0x1F7		/* offset of number of tracks (word)*/
#define ELKS_BPB_SecPerTrk	0x1F9		/* offset of sectors per track (byte)*/
#define ELKS_BPB_NumHeads	0x1FA		/* offset of number of heads (byte)*/

/* MINIX-only offsets*/
#define MINIX_SectOffset	0x1F3		/* offset of partition start sector (long)*/

/* FAT BPB start and end offsets*/
#define FATBPB_START	11				/* start at bytes per sector*/
#define FATBPB_END		61				/* through end of file system type*/

/* FAT-only offsets */
#define FAT_BPB_RootEntCnt	0x11		/* offset of FAT12/16 root directory entry count (word) */
#define FAT_BPB_FATSz16		0x16		/* offset of FAT12/16 allocation table sector count (word) */
#define FAT_BPB_SecPerTrk	0x18		/* offset of sectors per track (word) */
#define FAT_BPB_NumHeads	0x1A		/* offset of number of heads (word) */
#define FAT_BPB_SectOffset	0x1C		/* offset of partition start sector (long) */
#define FAT_BPB_RootClus	0x2C		/* offset of FAT32 root directory start cluster (long) */

#define PARTITION_START		0x01be		/* offset of partition table in MBR*/
#define PARTITION_END		0x01fd		/* end of partition 4 in MBR*/

unsigned int SecPerTrk, NumHeads, NumTracks;
unsigned long Start_sector;

int bootsecsize;
int rootfstype;
char bootblock[1024];				/* 1024 for MINIX, 512 for FAT */
char *rootdevice;
char *fsname[3] = { "Unknown", "Minix", "FAT" };

/* return /dev name of device*/
char *devname(dev_t dev)
{
	DIR *dp;
	struct dirent *d;
	struct stat st;
	static char devdir[] = "/dev";
	static char name[16];

	dp = opendir(devdir);
	if (dp == 0) {
		perror(devdir);
		return NULL;
	}
	strcpy(name, devdir);
	strcat(name, "/");

	while ((d = readdir(dp)) != NULL) {
		if (d->d_name[0] == '.')
 			continue;
		strcpy(name + sizeof(devdir), d->d_name);
		if (stat(name, &st) == 0) {
			if (S_ISBLK(st.st_mode) && st.st_rdev == dev) {
				closedir(dp);
				return name;
			}
		}
	}
	closedir(dp);
	fprintf(stderr, "Can't find device: 0x%x\n", dev);
	return NULL;
}

/* determine and return filesystem type*/
int get_fstype(int fd)
{
	struct minix_super_block *sb;
	char superblock[512];

	lseek(fd, 1024L, SEEK_SET);
	if (read(fd, superblock, 512) != 512) {
		fprintf(stderr, "Can't read superblock\n");
		return 0;
	}
	sb = (struct minix_super_block *)superblock;
	if (sb->s_magic == MINIX_SUPER_MAGIC)
		return FST_MINIX;
	return FST_MSDOS;		/* guess FAT if not MINIX*/
}

/* get device geometry*/
int get_geometry(int fd)
{
	struct hd_geometry geom;

	if (ioctl(fd, HDIO_GETGEO, &geom) < 0) {
		fprintf(stderr, "Can't get boot drive geometry\n");
		return 0;
	}

	NumTracks = geom.cylinders;
	NumHeads = geom.heads;
	SecPerTrk = geom.sectors;
	Start_sector = geom.start;

	return 1;	// success
}

/* set ELKS EPB track, head, sector and start_offset values in buffer*/
void setEPBparms(char *buf)
{
	buf[ELKS_BPB_SecPerTrk] = (unsigned char)SecPerTrk;
	buf[ELKS_BPB_NumHeads] = (unsigned char)NumHeads;
	buf[ELKS_BPB_NumTracks] = (unsigned char)NumTracks;
	buf[ELKS_BPB_NumTracks+1] = (unsigned char)(NumTracks >> 8);
}

/* set MINIX SectorOffset in buffer*/
void setMINIXparms(char *buf)
{
	*(unsigned long *)&buf[MINIX_SectOffset] = Start_sector;
}

/* set FAT BPB values from target in buffer*/
int setFATparms(int fd, char *buf)
{
	int n;
	char BPB[512];

	lseek(fd, 0L, SEEK_SET);
	n = read(fd, BPB, 512);
	if (n != 512) {
		fprintf(stderr, "Can't read target boot sector\n");
		return 0;
	}

	/* copy FAT BPB*/
	for (n = FATBPB_START; n <= FATBPB_END; n++)
		buf[n] = BPB[n];

	printf("BPB sect_offset %ld setting to %ld\n",
		*(unsigned long *)&buf[FAT_BPB_SectOffset], Start_sector);
#if 1
	*(unsigned short *)&buf[FAT_BPB_SecPerTrk] = SecPerTrk;
	*(unsigned short *)&buf[FAT_BPB_NumHeads] = NumHeads;
	*(unsigned long *)&buf[FAT_BPB_SectOffset] = Start_sector;
#endif
	return 1;
}

/* write MBR code outside of partition table*/
int setMBRparms(int fd)
{
	int n;
	char MBR[512];
	unsigned short *bsect = (unsigned short *)MBR;

	lseek(fd, 0L, SEEK_SET);
	n = read(fd, MBR, 512);
	if (n != 512) {
		fprintf(stderr, "Can't read target MBR\n");
		return 0;
	}

	/* copy MBR code*/
	for (n = 0; n < 512; n++)
		if (n < PARTITION_START || n > PARTITION_END)
			MBR[n] = mbr_bin[n];
	if (bsect[255] != 0xaa55)
		fprintf(stderr, "Warning: No boot signature in MBR\n");
	lseek(fd, 0L, SEEK_SET);
	n = write(fd, MBR, 512);
	if (n != 512) {
		fprintf(stderr, "Can't write target MBR\n");
		return 0;
	}
	return 1;
}

int copyfile(char *srcname, char *destname, int setmodes)
{
	int		rfd;
	int		wfd;
	int		rcc;
	int		wcc;
	char		*bp;
	struct	stat	statbuf1;
	struct	stat	statbuf2;
	struct	utimbuf	times;
	static	char	buf[BUF_SIZE];

	printf("Copying %s to %s\n", srcname, destname);
	if (stat(srcname, &statbuf1) < 0) {
		perror(srcname);
		return 0;
	}

	if (stat(destname, &statbuf2) < 0) {
		statbuf2.st_ino = -1;
		statbuf2.st_dev = -1;
	}

	rfd = open(srcname, 0);
	if (rfd < 0) {
		perror(srcname);
		return 0;
	}

	wfd = creat(destname, statbuf1.st_mode);
	if (wfd < 0) {
		perror(destname);
		close(rfd);
		return 0;
	}

	while ((rcc = read(rfd, buf, BUF_SIZE)) > 0) {
		bp = buf;
		while (rcc > 0) {
			wcc = write(wfd, bp, rcc);
			if (wcc < 0) {
				perror(destname);
				goto error_exit;
			}
			bp += wcc;
			rcc -= wcc;
		}
	}

	if (rcc < 0) {
		perror(srcname);
		goto error_exit;
	}

	close(rfd);
	if (close(wfd) < 0) {
		perror(destname);
		return 0;
	}

	if (setmodes) {
		chmod(destname, statbuf1.st_mode);
		chown(destname, statbuf1.st_uid, statbuf1.st_gid);
		times.actime =	statbuf1.st_atime;
		times.modtime = statbuf1.st_mtime;
		utime(destname, &times);
	}

	return 1;	// success

error_exit:
	close(rfd);
	close(wfd);

	return 0;	// fail
}

void fatalmsg(const char *s, ...)
{
	va_list p;

	va_start(p, s);
	fprintf(stderr, "Error: ");
	vfprintf(stderr, s, p);
	va_end(p);
	exit(-1);
}

/*
 * Read bootblock from current partition or from file.
 */
void get_bootblock(char *bootf) 
{
	int fd, n;
	unsigned short *bsect = (unsigned short *)bootblock;
	
	if (bootf == NULL) {	/* get bootblock from current root */
		fd = open(rootdevice, O_RDONLY);
		if (fd < 0)
			fatalmsg("Can't open boot device %s\n", rootdevice);

		if (!get_geometry(fd))
			fatalmsg("Can't get geometry for boot device %s\n", rootdevice);
		rootfstype = get_fstype(fd);
		if (rootfstype == 0)
			fatalmsg("Unknown boot device filesystem\n");
		printf("System on %s: %s (CHS %d/%d/%d at offset %ld)\n",
			rootdevice, fsname[rootfstype], NumTracks, NumHeads, SecPerTrk, Start_sector);
		bootsecsize = (rootfstype == FST_MINIX)? 1024: 512;

	} else {	/* get bootblock from file */

		fd = open(bootf, O_RDONLY);
		if (fd < 0)
			fatalmsg("Can't open boot file %s\n", bootf);
		bootsecsize = 1024;
	}
	lseek(fd, 0L, SEEK_SET);
	n = read(fd, bootblock, bootsecsize);
	if (n == 512) { rootfstype = FST_MSDOS; }
	else if (n == 1024) { rootfstype = FST_MINIX; }
	else { fatalmsg("Bootblock size error %d\n", n); }

	if (bsect[255] != 0xaa55)
		fatalmsg("Bad signature in bootblock\n");
	close(fd);
}

int main(int ac, char **av)
{
	char *targetdevice;
	char *bootfile = 0;
	int fstype = -1, fd, n;
	int opt_writembr = 0;
	int opt_writeflat = 0;
	int opt_syscopy = 0;
	int opt_writebb = 0;
	int fat32 = 0;
	dev_t rootdev, targetdev;
	struct stat sbuf;

	if (ac < 2 || ac > 5) {
usage:
		fatalmsg("Usage: makeboot [-M][-F] [-f file] [-s] /dev/{fd0,fd1,hda1,hda2,etc}\n");
	}
	if (ac == 2) opt_writebb++;
	while (av[1] && av[1][0] == '-') {
		if (av[1][1] == 'M') {
			opt_writembr = 1;
			av++;
			ac--;
		}
		else if (av[1][1] == 'F') {
			opt_writeflat = 1;
			av++;
			ac--;
		}
		else if (av[1][1] == 's') {
			opt_syscopy = 1;
			opt_writebb = 1;
			av++;
			ac--;
		}
		else if (av[1][1] == 'f') {
			av++;
			ac--;
			opt_writebb = 1;
			bootfile = av[1];
			av++;
			ac--;
		} else goto usage;
	}
	if (ac != 2)
		goto usage;
	targetdevice = av[1];

	if (stat("/", &sbuf) < 0) {
		perror("/");
		return -1;
	}

	rootdev = sbuf.st_dev;
	rootdevice = devname(rootdev);

	if (opt_writebb == 1) {
		get_bootblock(bootfile);
		fstype = rootfstype;
	}

	fd = open(targetdevice, O_RDWR);
	if (fd < 0)
		fatalmsg("Can't open target device %s\n", targetdevice);

	/* check target is not boot device, raw device, or -M used with floppy*/
	if (fstat(fd, &sbuf) < 0)
		fatalmsg("Can't stat %s\n", targetdevice);
	targetdev = sbuf.st_rdev;

	if (rootdev == targetdev)	/* FIXME: Should be able to MBR rootdev */
		fatalmsg("Can't specify current boot device as target\n");

	if (MINOR(targetdev) < BIOS_FD0_MINOR) {	/* hard drive*/
		if (!opt_writeflat && !opt_writembr) {
			if ((targetdev & BIOS_MINOR_MASK) == 0)	/* non-partitioned device*/
				fatalmsg("Must specify partitioned device (example /dev/hda1)\n");
		}
	} else {					/* floppy*/
		if (opt_writembr)
			fatalmsg("Can't use -M on floppy devices\n");
	}

	/* Write MBR */
	if (opt_writembr) {
		int ffd;

		char *rawtargetdevice = devname(targetdev & ~BIOS_MINOR_MASK);
		if (!rawtargetdevice)
			fatalmsg("Can't find raw target device\n");
		ffd = open(rawtargetdevice, O_RDWR);
		if (ffd < 0)
			fatalmsg("Can't open raw target device %s\n", targetdevice);
		if (setMBRparms(ffd))
			printf("Writing MBR on %s\n", rawtargetdevice);
		close(ffd);
	}

	if (opt_writeflat || opt_writebb) {
		if (!get_geometry(fd))		/* gets ELKS CHS parameters for bootblock write*/
			fatalmsg("Can't get geometry for target device %s\n", targetdevice);
		setEPBparms(bootblock);		/* sets ELKS CHS parameters in bootblock*/
		fstype = get_fstype(fd);
		printf("Target on %s: %s (CHS %d/%d/%d at offset %ld)\n",
			targetdevice, fsname[fstype], NumTracks, NumHeads, SecPerTrk, Start_sector);

		if (rootfstype != fstype)
			fatalmsg("Target and System filesystem must be same type\n");

		if (fstype == FST_MINIX) setMINIXparms(bootblock);
		if (fstype == FST_MSDOS) {
			if (!setFATparms(fd, bootblock)) {
				close(fd);
				return -1;
			}
			fat32 = (*(unsigned short *)&bootblock[FAT_BPB_FATSz16] == 0);
			if (fat32) {	/* if FAT32, check if root directory starts on cluster 2 */
				unsigned long rootclus;
				rootclus = *(unsigned long *)&bootblock[FAT_BPB_RootClus];
				if (rootclus != 2)
				    fatalmsg("Unsupported: weird root directory start cluster (%lu)\n",
						rootclus);
			}
		}

		lseek(fd, 0L, SEEK_SET);
		n = write(fd, bootblock, bootsecsize);
		if (n != bootsecsize)
			fatalmsg("Can't write target boot sector\n");
		close(fd);
		printf("Bootblock written\n");
	}

	if (opt_syscopy == 1) {
		if (mkdir(MOUNTDIR, 0777) < 0)
			fprintf(stderr, "Temp mount point %s not created (may already exist)\n", MOUNTDIR);

		if (mount(targetdevice, MOUNTDIR, fstype, 0) < 0) {
			fprintf(stderr, "Error: Can't mount %s on %s\n", targetdevice, MOUNTDIR);
			goto errout2;
		}
		if (!copyfile(SYSFILE1, MOUNTDIR SYSFILE1, 1)) {
			fprintf(stderr, "Error copying %s\n", SYSFILE1);
			goto errout;
		}

		if (fstype == FST_MSDOS) {
			if (mkdir(MOUNTDIR DEVDIR, 0777) < 0)
				fprintf(stderr, "/dev directory may already exist on target\n");
		} else {
			if (!copyfile(SYSFILE2, MOUNTDIR SYSFILE2, 1))
				fprintf(stderr, "Not copying %s file\n", SYSFILE2);
		}
	
		if (umount(targetdevice) < 0)
			fprintf(stderr, "Unmount error\n");
		rmdir(MOUNTDIR);
		printf("System copied\n");
	}

	sync();

	return fstype;		/* return filesystem type (1=MINIX, 2=FAT) */

errout:
	umount(targetdevice);
errout2:
	rmdir(MOUNTDIR);
	sync();
	return -1;
}
